### ISSUE TITLE
Investigate Deploy to Heroku

### ISSUE THREAD

**Issue by ismisepaul** [OWASP/SecurityShepherd#504]
As a user of security shepherd
I want a quick and easy way to run the application in the cloud through a Heroku button in the Readme
because it should be simple to test out Security Shepherd

**Comment by ismisepaul** [comment]
**Deploying WAR**

1. Build the war
2. Push to heroku ```heroku deploy:war --war owaspSecurityShepherd.war -a secsheptest```

But it doesn't seem to generate the auth file;
```
2019-10-07T15:21:12.545715+00:00 app[web.1]: ERROR 10.32.165.53 from 149.11.36.150 @ 07 Oct 15:21:12,545 [http-nio-12738-exec-6] (FileInputProperties.java:81) - Error: Properties filename: /app/target/tomcat.12738//conf/database.properties can not be found: /app/target/tomcat.12738/conf/database.properties (No such file or directory)
2019-10-07T15:21:12.545989+00:00 app[web.1]: FATAL 10.32.165.53 from 149.11.36.150 @ 07 Oct 15:21:12,545 [http-nio-12738-exec-6] (Database.java:165) - Unable to create database connection: java.lang.ClassNotFoundException: NO RESULTjava.io.FileNotFoundException: /app/target/tomcat.12738/conf/database.properties (No such file or directory)
2019-10-07T15:21:12.546367+00:00 app[web.1]: FATAL 10.32.165.53 from 149.11.36.150 @ 07 Oct 15:21:12,546 [http-nio-12738-exec-6] (Setup.java:188) - Unable to generate auth
2019-10-07T15:21:12.546414+00:00 app[web.1]: java.nio.file.NoSuchFileException: /app/target/tomcat.12738/conf/SecurityShepherd.auth
```

Running heroku bash that folder structure doesn't exists ```heroku run bash -a secsheptest```;
```Running bash on ⬢ secsheptest... up, run.7298 (Free)
~ $ ls /app/
Procfile  owaspSecurityShepherd.war  system.properties	webapp-runner.jar
~ $ 
```

Launching the web app;
![image](https://user-images.githubusercontent.com/912403/66326347-b4059200-e920-11e9-82f3-8bc5269ef1a9.png)

Tried creating that folder structure and then even tried creating the auth file. No luck. 

**Docker**
Pull the OWASP image 
```docker pull owasp/security-shepherd``` 
Tag and push the container
```heroku container:login```
```docker tag owasp/security-shepherd registry.heroku.com/secsheptest/web```
```docker push registry.heroku.com/secsheptest/web```
Release the container and run it on the app secsheptest
```heroku container:release web -a secsheptest```

The folder structure is there with ```conf```
```
Running bash on ⬢ secsheptest... up, run.7332 (Free)
~ $ ls
BUILDING.txt     LICENSE          README.md        RUNNING.txt      conf             lib              native-jni-lib   webapps
CONTRIBUTING.md  NOTICE           RELEASE-NOTES    bin              include          logs             temp             work
~ $ 
```

Launching the web app
![image](https://user-images.githubusercontent.com/912403/66327829-62aad200-e923-11e9-8635-8de31b8695c2.png)


Viewing the logs
```
2019-10-07T15:21:15.422156+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
```

Heroku doesn't respect the ```EXPOSE``` in the [dockerfile ](https://help.heroku.com/PPBPA231/how-do-i-use-the-port-environment-variable-in-container-based-apps) but instead it uses a dynamic port number on startup.

I was going down the rabbit hole of changing port 8080 and 8443 in ```server.xml``` to the dynamic port set by Heroku. But no luck. It seems like setting this up with Tomcat on Docker doesn't work - ref: https://medium.com/thinkspecial/heroku-traditional-war-vs-docker-deployment-analysis-e592816ef9d4

**Comment by ismisepaul** [comment]
Added env variable to get the producing of auth and database.properties file working, but still not working. Idea from here - https://github.com/heroku/webapp-runner/issues/57 Also added addons for mysql and mongodb. Commit - https://github.com/OWASP/SecurityShepherd/commit/831d4f5eb68f9a1646d9a1fdbaeff7d760767dbe 

Procfile uses the $WAR_DIR to set the dir to the expanded dir https://github.com/OWASP/SecurityShepherd/commit/5d0001b69179b5fff6fcacf3768bff56fc46e59f
```
2019-12-27T11:33:36.732706+00:00 app[web.1]: FATAL  @ 27 Dec 11:33:36,732 [http-nio-8305-exec-8] (Setup.java:247) - Database properties file not found, assuming not installed: java.io.FileNotFoundException: /app/target/tomcat.8305/conf/database.properties (No such file or directory)
2019-12-27T11:33:36.733609+00:00 app[web.1]: FATAL  @ 27 Dec 11:33:36,733 [http-nio-8305-exec-8] (Setup.java:286) - Unable to generate auth: /app/target/tomcat.8305/conf/SecurityShepherd.auth
2019-12-27T11:33:36.738468+00:00 app[web.1]: Dec 27, 2019 11:33:36 AM org.apache.catalina.core.StandardWrapperValve invoke
2019-12-27T11:33:36.738470+00:00 app[web.1]: SEVERE: Servlet.service() for servlet [default] in context with path [] threw exception
2019-12-27T11:33:36.738471+00:00 app[web.1]: java.lang.RuntimeException: java.nio.file.NoSuchFileException: /app/target/tomcat.8305/conf/SecurityShepherd.auth
2019-12-27T11:33:36.738473+00:00 app[web.1]: 	at servlets.Setup.generateAuth(Setup.java:287)
```

When you run `heroku run bash -a <app-name>` the dir doesn't exist nor does the `app/target/webapp`

**Comment by jules2689** [comment]
Hey, so I took a look at this and was able to get as far as you using your code. I suspect the issue here is that each dyno and even the `run bash` dyno that you get is entirely different. They aren't identical.

When Heroku starts a tomcat server, it sets up in that folder. So in the dyno you're running on, this folder exists. I suspect `8305` in `/app/target/tomcat.8305/conf/SecurityShepherd.auth` is the PID of the tomcat server.

Looking at SecurityShepard's code, I suspect that what needs to change is this line: https://github.com/OWASP/SecurityShepherd/blob/dev/src/main/java/dbProcs/Constants.java#L6

This is what is used to get that path.

> Added env variable to get the producing of auth and database.properties file working, but still not working.

I couldn't find a produced `database.properties` file, but I suspect if we can change the `cataline.base` or even allow it to be overidden by an ENV var, we may be able to get those files found.

**Comment by jules2689** [comment]
So I hand modified the constant to be `/app/target`, and set the web startup to be a `run.sh` file that creates the `conf` folder (hacky to test it out).

This leaves us with a `18:49:36,044 [http-nio-49330-exec-8] (Database.java:165) - Unable to create database connection: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure`

<details>
  
```
2020-01-01T18:49:35.931764+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,930 [http-nio-49330-exec-4] (Setup.java:176) - Heroku database properties written to: /app/target/conf/database.properties
2020-01-01T18:49:35.932164+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,931 [http-nio-49330-exec-4] (Setup.java:177) - Files in conf: java.util.stream.ReferencePipeline$Head@4ba79b79
2020-01-01T18:49:35.932530+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,931 [http-nio-49330-exec-7] (Setup.java:176) - Heroku database properties written to: /app/target/conf/database.properties
2020-01-01T18:49:35.932965+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,932 [http-nio-49330-exec-7] (Setup.java:177) - Files in conf: java.util.stream.ReferencePipeline$Head@dbe5221
2020-01-01T18:49:35.933695+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,932 [http-nio-49330-exec-5] (Setup.java:176) - Heroku database properties written to: /app/target/conf/database.properties
2020-01-01T18:49:35.934221+00:00 app[web.1]: INFO  @ 01 Jan 18:49:35,933 [http-nio-49330-exec-5] (Setup.java:177) - Files in conf: java.util.stream.ReferencePipeline$Head@608b37bd
2020-01-01T18:49:36.044964+00:00 app[web.1]: FATAL  @ 01 Jan 18:49:36,044 [http-nio-49330-exec-8] (Database.java:165) - Unable to create database connection: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure
```

</details>

With this in mind, I logged what we were putting into the `database.properties` file. I was surprised that port was `-1`. When I looked at the `MYSQL_URL` from ClearDB, they specify no port. So I manually injected 3306.

This brought me to a "working state" in that we can now connect to the DB.

There is one issue now: `com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: User 'foobar' has exceeded the 'max_user_connections' resource (current value: 10)`.... seems we have some connections issues. This is preventing anything from completing

**Comment by jules2689** [comment]
Hacky Diff to get to that state:
```diff
diff --git a/Procfile b/Procfile
index fb0ed1fe..0a6c6a57 100644
--- a/Procfile
+++ b/Procfile
@@ -1 +1 @@
-web: java $JAVA_OPTS -jar target/dependency/webapp-runner.jar --expanded-dir-name $WAR_DIR --port $PORT target/*.war
+web: bash run.sh
diff --git a/run.sh b/run.sh
new file mode 100755
index 00000000..2e81e294
--- /dev/null
+++ b/run.sh
@@ -0,0 +1,4 @@
+mkdir -p /app/target
+mkdir -p /app/target/conf
+
+java $JAVA_OPTS -jar target/dependency/webapp-runner.jar --expanded-dir-name $WAR_DIR --port $PORT target/*.war
diff --git a/src/main/java/dbProcs/Constants.java b/src/main/java/dbProcs/Constants.java
index d405eb87..19fc2488 100644
--- a/src/main/java/dbProcs/Constants.java
+++ b/src/main/java/dbProcs/Constants.java
@@ -3,8 +3,8 @@ package dbProcs;
 import java.io.File;

 public final class Constants {
-	public static final String CATALINA_BASE = System.getProperty("catalina.base");
-	public static final String CATALINA_CONF = CATALINA_BASE + File.separator + File.separator + "conf";
+	public static final String CATALINA_BASE = "/app/target";
+	public static final String CATALINA_CONF = CATALINA_BASE + File.separator + "conf";
 	public static final String DBPROP = CATALINA_CONF + File.separator + "database.properties";
 	public static final String SETUP_AUTH = CATALINA_CONF + File.separator + "SecurityShepherd.auth";
 	public static final String MONGO_DB_PROP = CATALINA_CONF + File.separator + "mongo.properties";
diff --git a/src/main/java/servlets/Setup.java b/src/main/java/servlets/Setup.java
index a6eabc87..d6387167 100644
--- a/src/main/java/servlets/Setup.java
+++ b/src/main/java/servlets/Setup.java
@@ -318,7 +318,7 @@ public class Setup extends HttpServlet {
 		String dbPass = dbUri.getUserInfo().split(":")[1];

 		StringBuffer dbProp = new StringBuffer();
-		dbProp.append("databaseConnectionURL=jdbc:mysql://" + dbUri.getHost() + ':' + dbUri.getPort() + dbUri.getPath());
+		dbProp.append("databaseConnectionURL=jdbc:mysql://" + dbUri.getHost() + ":3306" + dbUri.getPath());
 		dbProp.append("\n");
 		dbProp.append("DriverType=org.gjt.mm.mysql.Driver");
 		dbProp.append("\n");
```

**Comment by jules2689** [comment]
I found a number of getters for MySQL that open connections but don't seem to close them in `Getter.java`, I tried to just add those in, but that didn't seem to help. There's something leaking connections somewhere because I end up with a bunch of sleeping connections in the ClearDB console:

![image](https://user-images.githubusercontent.com/3074765/71645154-eade1080-2ca1-11ea-8672-33de796ac485.png)


Edit: I added a ton of `finally` blocks with `Database.closeConnection(conn)` and that seems to have worked! There's a leak somewhere in there, I suspect, due to an early return in a try/catch

Now it's trying to connect to `<database_name>core`... this is the DB Schema I think. Removing the schema did seem to help.

I also got errors with `PROCEDURE`s:

```
2020-01-01T19:25:03.146974+00:00 app[web.1]: FATAL 10.45.208.208 from 69.159.41.49 @ 01 Jan 19:25:03,146 [http-nio-38506-exec-9] (Setter.java:993) - userCreate Failure: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: PROCEDURE heroku_c2fd4f21c18f103.userCreate does not exist
2020-01-01T19:25:03.148125+00:00 app[web.1]: ERROR 10.45.208.208 from 69.159.41.49 @ 01 Jan 19:25:03,147 [http-nio-38506-exec-9] (Register.java:160) - Registration Error: java.sql.SQLException: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: PROCEDURE heroku_c2fd4f21c18f103.userCreate does not exist
```

**Comment by ismisepaul** [comment]
I was able to get this partially working where I;

1. Created the dir structure (this allows us to keep the catalina.base constant) https://github.com/OWASP/SecurityShepherd/blob/a9b55c19304009fc66e4af0b657bca05e3c13377/src/main/java/servlets/Setup.java#L360
2. If a Heroku env is detected then it creates the DB props

The problem is you don't get to choose the name of your DB. Security Shepherd users "core" and is hardcoded in places. However, when a cleardb is setup up with Heroku it creates this random db name. So I replace all the instances of "core" with the name of the cleardb database;
https://github.com/OWASP/SecurityShepherd/blob/a9b55c19304009fc66e4af0b657bca05e3c13377/src/main/java/servlets/Setup.java#L303

The I also remove the last part where a backup db is created as creating schemas doesn't seem to work with ClearDb on Heroku  
https://github.com/OWASP/SecurityShepherd/blob/a9b55c19304009fc66e4af0b657bca05e3c13377/src/main/java/servlets/Setup.java#L305

I am able to spin up an instance of Security Shepherd using the Heroku button (I don't know how to get this working for a branch) But it does work here on my fork https://github.com/ismisepaul/SecurityShepherd/tree/master

The problem is with the level schemas. I don't really know enough but I think that a DB will need to be created for each (ClearDB) This is possible on Heroku by specifing them within the app.json https://github.com/OWASP/SecurityShepherd/blob/a9b55c19304009fc66e4af0b657bca05e3c13377/app.json#L6

@jules2689 thanks for looking into this and we're aware of the problems with exceptions and db connections staying open. I think @etnoy has done a lot of work in that area and we're just waiting on the pull request to be merged in. Take a look at my latest commit https://github.com/OWASP/SecurityShepherd/commit/a9b55c19304009fc66e4af0b657bca05e3c13377 or my fork it's up and running but without the levels that rely on a db.

I think there's a lot more work in this to get it working with Heroku.

**Comment by jules2689** [comment]
That commit looks like it'll do what my hacky `run.sh` did! Nice 😄 
This line seems to have changed from what I checked out as well (https://github.com/OWASP/SecurityShepherd/blob/a9b55c19304009fc66e4af0b657bca05e3c13377/src/main/java/servlets/Setup.java#L402). It used to say `=core`... so that should fix the other issue I encountered. The port is gone now too 😃 

At this point, it feels less like there's specific work on Heroku (as the deploy to Heroku button worked for me too) and we're waiting on the changes with `PROCEDURE` and open connections.

This is looking fantastic. I'm hoping to use this in a software course I teach at a local university. One of their assignments will be to complete `n` of the challenges, and Heroku makes it easy enoguh to host that I don't have to worry about issues there.

Thanks for your work ❤️
